Extension cases sync

proteusvacuum opened the pull request:
Ready for review! I tried to clean up the commits so you can most likely go through :blowfish: / :blowfish: 

I haven't added the performance enhancement we talked about last week, as I was still putting the finishing touches on here. Will do that in a separate PR so that this one doesn't grow any larger than it already is.

@ctsims @czue @dannyroberts 
code buddy: @benrudolph 
spec: https://gist.github.com/dannyroberts/f184daad468fb7debf10

proteusvacuum committed:
Emit relationship from case/related view

Update dbaccessors + tests to take advantage of this

proteusvacuum committed:
New purge algorithm

proteusvacuum committed:
Update phone lists with extension indices

proteusvacuum committed:
Purge cases during clean ownership sync

proteusvacuum committed:
Tests for purge algorithm

proteusvacuum committed:
Tests for sync

proteusvacuum committed:
Update dirtiness flag on sync

proteusvacuum committed:
Get extension cases in footprint

proteusvacuum committed:
Get cleanliness from scratch with extensions

proteusvacuum committed:
Pull whole footprint on first extension sync

proteusvacuum committed:
Only fetch extensions when they are wanted, mostly

proteusvacuum committed:
Sync many times

dimagimon commented on corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py:
```diff
@@ -20,6 +22,7 @@
 from casexml.apps.phone.restore import CachedResponse, RestoreConfig, RestoreParams, RestoreCacheSettings
 from casexml.apps.case.xml import V2, V1
 from casexml.apps.case.sharedmodels import CommCareCaseIndex
+from casexml.apps.case.const import CASE_INDEX_EXTENSION
```
F401 'CASE_INDEX_EXTENSION' imported but unused

ctsims replied:
Should we synchronize on our test plans? the one at

https://github.com/dimagi/commcare/blob/master/tests/resources/case_relationship_tests.json

is up to date with our fixes.

On Fri, Nov 6, 2015 at 4:48 PM, dimagimon <notifications@github.com> wrote:

> In corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py
> <https://github.com/dimagi/commcare-hq/pull/9012#discussion_r44191422>:
>
> > @@ -20,6 +22,7 @@
> >  from casexml.apps.phone.restore import CachedResponse, RestoreConfig, RestoreParams, RestoreCacheSettings
> >  from casexml.apps.case.xml import V2, V1
> >  from casexml.apps.case.sharedmodels import CommCareCaseIndex
> > +from casexml.apps.case.const import CASE_INDEX_EXTENSION
>
> F401 'CASE_INDEX_EXTENSION' imported but unused
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/9012/files#r44191422>.
>

proteusvacuum replied:
@ctsims, yes. Will merge your changes with mine, (also added a bunch of tests), then submit a PR to your repo.

proteusvacuum replied:
One major change is that I made all the names of the tests snake_case so that they would somewhat pass the grep test whenever a test failed.

proteusvacuum committed:
lint

czue commented:
wow, my head hurts after reviewing this. nice work!

czue commented:
(also assume you'll look at test failures, which look related)

```

======================================================================
ERROR: test_close_host (casexml.apps.phone.tests.test_sync_mode.ExtensionCasesSyncTokenUpdates)
closing a host should update the appropriate trees
----------------------------------------------------------------------
Traceback (most recent call last):
  File "corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py", line 860, in test_close_host
    self.factory.create_or_update_cases([extension])
  File "corehq/ex-submodules/casexml/apps/case/mock.py", line 333, in create_or_update_cases
    form_extras,
  File "corehq/ex-submodules/casexml/apps/case/mock.py", line 297, in post_case_blocks
    domain=self.domain,
  File "/home/travis/build/dimagi/commcare-hq/corehq/form_processor/interfaces/processor.py", line 79, in post_case_blocks
    return post_case_blocks(case_blocks, form_extras=form_extras, domain=domain)
  File "corehq/ex-submodules/casexml/apps/case/util.py", line 48, in post_case_blocks
    response, xform, cases = submit_form_locally(form, domain, **form_extras)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/receiverwrapper/util.py", line 24, in submit_form_locally
    **kwargs
  File "corehq/ex-submodules/couchforms/util.py", line 374, in run
    instance, cases, errors = self.process_xforms_for_cases(lock_manager)
  File "corehq/ex-submodules/couchforms/util.py", line 395, in process_xforms_for_cases
    case_result = self.interface.process_cases_with_casedb(xforms, case_db)
  File "/home/travis/build/dimagi/commcare-hq/corehq/form_processor/interfaces/processor.py", line 97, in process_cases_with_casedb
    return self.processor.process_cases_with_casedb(xforms, case_db)
  File "/home/travis/build/dimagi/commcare-hq/corehq/form_processor/backends/couch/processor.py", line 89, in process_cases_with_casedb
    return process_cases_with_casedb(xforms, case_db)
  File "corehq/ex-submodules/casexml/apps/case/xform.py", line 100, in process_cases_with_casedb
    case_processing_result = _get_or_update_cases(xforms, case_db)
  File "corehq/ex-submodules/casexml/apps/case/xform.py", line 272, in _get_or_update_cases
    dirtiness_flags += list(_get_dirtiness_flags_for_outgoing_indices(case))
  File "corehq/ex-submodules/casexml/apps/case/xform.py", line 238, in _get_dirtiness_flags_for_outgoing_indices
    for index in case.indices if case_db.doc_exist(index.referenced_id)} | tree_owners
  File "corehq/ex-submodules/casexml/apps/case/xform.py", line 237, in <setcomp>
    owner_ids = {case_db.get(index.referenced_id).owner_id
AttributeError: 'CaseDbCacheCouch' object has no attribute 'doc_exist'
```

```
======================================================================
ERROR: testLotsOfSubcases (casexml.apps.case.tests.test_bugs.CaseBugTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "corehq/ex-submodules/casexml/apps/case/tests/test_bugs.py", line 126, in testLotsOfSubcases
    response, form, cases = self.interface.submit_form_locally(xml_data)
  File "/home/travis/build/dimagi/commcare-hq/corehq/form_processor/interfaces/processor.py", line 76, in submit_form_locally
    return submit_form_locally(instance, domain, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/receiverwrapper/util.py", line 24, in submit_form_locally
    **kwargs
  File "corehq/ex-submodules/couchforms/util.py", line 374, in run
    instance, cases, errors = self.process_xforms_for_cases(lock_manager)
  File "corehq/ex-submodules/couchforms/util.py", line 395, in process_xforms_for_cases
    case_result = self.interface.process_cases_with_casedb(xforms, case_db)
  File "/home/travis/build/dimagi/commcare-hq/corehq/form_processor/interfaces/processor.py", line 97, in process_cases_with_casedb
    return self.processor.process_cases_with_casedb(xforms, case_db)
  File "/home/travis/build/dimagi/commcare-hq/corehq/form_processor/backends/couch/processor.py", line 89, in process_cases_with_casedb
    return process_cases_with_casedb(xforms, case_db)
  File "corehq/ex-submodules/casexml/apps/case/xform.py", line 100, in process_cases_with_casedb
    case_processing_result = _get_or_update_cases(xforms, case_db)
  File "corehq/ex-submodules/casexml/apps/case/xform.py", line 272, in _get_or_update_cases
    dirtiness_flags += list(_get_dirtiness_flags_for_outgoing_indices(case))
  File "corehq/ex-submodules/casexml/apps/case/xform.py", line 238, in _get_dirtiness_flags_for_outgoing_indices
    for index in case.indices if case_db.doc_exist(index.referenced_id)} | tree_owners
  File "corehq/ex-submodules/casexml/apps/case/xform.py", line 237, in <setcomp>
    owner_ids = {case_db.get(index.referenced_id).owner_id
AttributeError: 'CaseDbCacheCouch' object has no attribute 'doc_exist'
----------------------------------------------------------------------
```

czue commented on corehq/ex-submodules/casexml/apps/case/const.py:
```diff
@@ -52,3 +51,5 @@
 
 DEFAULT_CASE_INDEX_IDENTIFIERS = {CASE_INDEX_CHILD: 'parent', CASE_INDEX_EXTENSION: 'host'}
+
+UNOWNED_EXTENSION_OWNER_ID = '-'
```
Oh we should also add this to `WEIRD_USER_IDS` so we don't try to [determine cleanliness on it](https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/phone/cleanliness.py#L63)....

benrudolph commented:
nice dude

proteusvacuum committed:
Merge branch 'master' into extension-cases-sync-cleaned

proteusvacuum committed:
Update get dirtiness flags with new CouchDbCache

New model doesn't have `doc_exist`. Better to use `get` anyway to
populate the cache if we are going to use it.

proteusvacuum committed:
Merge tests from mobile

proteusvacuum committed:
Move design doc to couchapps

proteusvacuum committed:
Move open_cases to couchapps

proteusvacuum committed:
Move get_lite view to couchapps

proteusvacuum committed:
move form_case_index to couchapps

proteusvacuum committed:
Update misleading test, add another test

proteusvacuum committed:
Properly return json for legacy indices

proteusvacuum committed:
Update docstrings

proteusvacuum committed:
this_case_index -> referenced case

proteusvacuum committed:
Remove check for old synclogs not deleting properly

proteusvacuum committed:
Refactor get_dependent_case_info

Uses get_direct_dependencies in the loop to make it clearer

proteusvacuum committed:
Prune -> Purge everywhere

Also remove unused PruneResult

proteusvacuum committed:
Remove unused constants

proteusvacuum committed:
Don't process host cases multiple times

proteusvacuum committed:
Update docstrings in test_cleanliness

proteusvacuum committed:
Use a context manager instead of multiple tests

proteusvacuum committed:
rename case_ids_for_ variables to be clearer

proteusvacuum committed:
Update variable name so it doesn't look boolean

proteusvacuum committed:
Update comment to be right

proteusvacuum committed:
Add unowned extension owner id to weird users

dimagimon commented on corehq/couchapps/case_indices/views/related/map.js:
```diff
@@ -0,0 +1,23 @@
+function(doc) {
```
Missing name in function declaration.

proteusvacuum committed:
Add extensions checked as explicit property

proteusvacuum committed:
Merge branch 'master' into extension-cases-sync-cleaned

proteusvacuum committed:
Validate case removal

proteusvacuum committed:
Merge branch 'master' into extension-cases-sync-cleaned

proteusvacuum committed:
Include relationship on forwards index

dimagimon commented on corehq/couchapps/case_indices/views/related/map.js:
```diff
@@ -0,0 +1,30 @@
+                emit([doc.domain, doc._id, "index"], index);
+
+                var reverse_index = {};
+                for (var key in doc.indices[i]){
```
'key' is already defined.

proteusvacuum committed:
Allow closing multiple cases at once

proteusvacuum committed:
Get extension chain dbaccessor

proteusvacuum committed:
Close extensions when when host is closed

dimagimon commented on corehq/ex-submodules/casexml/apps/case/tests/test_close_extension_chain.py:
```diff
@@ -0,0 +1,102 @@
+from casexml.apps.phone.models import User
+from casexml.apps.phone.tests.test_sync_mode import SyncBaseTest
+from corehq.apps.domain.models import Domain
+from corehq.form_processor.interfaces.processor import FormProcessorInterface
```
F401 'FormProcessorInterface' imported but unused

proteusvacuum committed:
lint

proteusvacuum commented:
Commits 413d6b4, 572f96d, 18fdb65 close extensions when their host is closed.

proteusvacuum commented:
QA ticket: http://manage.dimagi.com/default.asp?187037#1045161

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into extension-cases-sync-cleaned

proteusvacuum committed:
Merge branch 'master' into extension-cases-sync-cleaned

dimagimon commented on corehq/ex-submodules/casexml/apps/phone/cleanliness.py:
```diff
@@ -13,10 +14,13 @@
 from corehq.apps.users.util import WEIRD_USER_IDS
 from django.conf import settings
 from corehq.util.soft_assert import soft_assert
+from dimagi.utils.couch.database import get_db, iter_docs
```
F401 'get_db' imported but unused

proteusvacuum commented:
Passed QA, G2M!

benrudolph commented:
:balloon: :balloon: woowoo!! :balloon: :balloon:

czue commented:
@dannyroberts do you think we should wait until next weekend to merge given the reindex?

dannyroberts commented:
Probably!

dannyroberts commented:
@proteusvacuum volunteered to merge and preindex at the end of the week while most of Boston is off for thanksgiving. Thanks!

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into extension-cases-sync-cleaned

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into extension-cases-sync-cleaned

proteusvacuum committed:
Close extensions

- fixes bad merge

proteusvacuum committed:
Remove unused import

czue commented:
:balloon: :balloon: :balloon: :balloon: :balloon: :balloon: :balloon: :balloon: :balloon:

proteusvacuum commented:
hooray!

snopoke commented:
✴✳✨🌟❇

benrudolph commented:
:briefcase: :link: :link: :link: :link: :briefcase:
